src/lib.rs

```diff
@@ -173,164 +194,388 @@ impl<const OSC_RAW_BUF_SIZE: usize> Parser<OSC_RAW_BUF_SIZE> {
+            _ => self.anywhere(performer, byte),
+        }
+    }
+
+    #[inline(always)]
+    fn advance_csi_param<P: Perform>(&mut self, performer: &mut P, byte: u8) {
+        match byte {
+            0x00..=0x17 | 0x19 | 0x1C..=0x1F => performer.execute(byte),
+            0x20..=0x2F => {
+                self.action_collect(byte);
+                self.state = State::CsiIntermediate
             },
-            _ => (),
+            0x30..=0x39 => self.action_paramnext(byte),
+            0x3A => self.action_subparam(),
+            0x3B => self.action_param(),
+            0x3C..=0x3F => self.state = State::CsiIgnore,
+            0x40..=0x7E => self.action_csi_dispatch(performer, byte),
+            0x7F => (),
+            _ => self.anywhere(performer, byte),
         }
+    }
 
-        if action == Action::None {
-            match state {
-                State::CsiEntry | State::DcsEntry | State::Escape => self.reset_params(),
-                State::DcsPassthrough => {
-                    if self.params.is_full() {
-                        self.ignoring = true;
-                    } else {
-                        self.params.push(self.param);
-                    }
+    #[inline(always)]
+    fn advance_dcs_entry<P: Perform>(&mut self, performer: &mut P, byte: u8) {
+        match byte {
+            0x00..=0x17 | 0x19 | 0x1C..=0x1F => (),
+            0x20..=0x2F => {
+                self.action_collect(byte);
+                self.state = State::DcsIntermediate
+            },
+            0x30..=0x39 => {
+                self.action_paramnext(byte);
+                self.state = State::DcsParam
+            },
+            0x3A => {
+                self.action_subparam();
+                self.state = State::DcsParam
+            },
+            0x3B => {
+                self.action_param();
+                self.state = State::DcsParam
+            },
+            0x3C..=0x3F => {
+                self.action_collect(byte);
+                self.state = State::DcsParam
+            },
+            0x40..=0x7E => self.action_hook(performer, byte),
+            0x7F => (),
+            _ => self.anywhere(performer, byte),
+        }
+    }
 
-                    performer.hook(
-                        self.params(),
-                        self.intermediates(),
-                        self.ignoring,
-                        byte as char,
-                    );
-                },
-                State::OscString => {
-                    self.osc_raw.clear();
-                    self.osc_num_params = 0;
-                },
-                _ => (),
-            }
-        } else {
-            self.perform_action(performer, action, byte);
+    #[inline(always)]
+    fn advance_dcs_ignore<P: Perform>(&mut self, performer: &mut P, byte: u8) {
+        match byte {
+            0x00..=0x17 | 0x19 | 0x1C..=0x7F => (),
+            _ => self.anywhere(performer, byte),
```

Comment:
Technically we could just replace this entire method body with a call to `self.anywhere`.